src/app/core/services/notifications/fcm-rest-notification.service.ts

```diff
@@ -0,0 +1,186 @@
+    this.initApiClient()
+  }
+
+  async initApiClient() {
+    await Swagger({ url: `${this.APP_SERVER_URL}/v3/api-docs` }).then(
+      client => {
+        this.apiClient = client
+      }
+    )
+  }
+
+  getSubjectDetails() {
+    return this.checkProjectAndSubjectExistElseCreate()
+  }
+
+  publishAllNotifications(user, sourceId, limit): Promise<any> {
+    return this.schedule.getTasks(TaskType.ALL).then(tasks => {
+      const fcmNotifications = this.notifications
+        .futureNotifications(tasks, limit)
+        .map(t => this.format(t, sourceId))
+      this.logger.log('NOTIFICATIONS Scheduling FCM notifications')
+      this.logger.log(fcmNotifications)
+      return Promise.all(
+        fcmNotifications
+          .map(n => this.sendNotification(n, user.subjectId, user.projectId))
+          .concat([this.setLastNotificationUpdate()])
+      )
+    })
+  }
+
+  publishTestNotification(user, sourceId): Promise<void> {
+    return this.sendNotification(
+      this.format(this.notifications.createTestNotification(), sourceId),
+      user.subjectId,
+      user.projectId
+    )
+  }
+
+  cancelAllNotifications(user): Promise<any> {
+    return this.apiClient.apis[
+      'fcm-notification-controller'
+    ].deleteNotificationsForUser({
+      subjectId: user.subjectId,
+      projectId: user.projectId
+    })
+  }
+
+  cancelSingleNotification(user, notificationId) {
+    return this.apiClient.apis['fcm-notification-controller']
+      .deleteNotificationUsingProjectIdAndSubjectIdAndNotificationId({
+        subjectId: user.subjectId,
+        projectId: user.projectId,
+        id: notificationId
+      })
+      .then(() =>
+        console.log('Success cancelling notification ' + notificationId)
+      )
+  }
+
+  private checkProjectAndSubjectExistElseCreate(): Promise<any> {
+    return this.checkProjectExistsElseCreate().then(() =>
+      this.checkSubjectExistsElseCreate()
+    )
+  }
+
+  private checkProjectExistsElseCreate(): Promise<any> {
+    return this.config.getProjectName().then(projectId => {
+      return this.apiClient.apis['radar-project-controller']
+        .getProjectsUsingProjectId({ projectId })
+        .catch(e => {
+          if (e.status == 404) {
+            return this.apiClient.apis['radar-project-controller'].addProject({
+              projectId
+            })
+          } else return Promise.reject(e)
+        })
+    })
+  }
+
+  private checkSubjectExistsElseCreate(): Promise<any> {
+    return Promise.all([
+      this.config.getEnrolmentDate(),
+      this.config.getProjectName(),
+      this.config.getParticipantLogin()
+    ]).then(([enrolmentDate, projectId, subjectId]) => {
+      return this.apiClient.apis['radar-user-controller']
+        .getRadarUserUsingSubjectId({ subjectId })
+        .then(res => res.body)
+        .catch(e =>
+          e.status == 404
+            ? this.apiClient.apis['radar-user-controller'].addUser({
```

Comment:
please use if/else if possible here. the `?:` is good for short expressions

Reply:
My bad, Bettercodehub was telling me to reduce the length of this function 😅